misc/metrics-extensions/modules/ballerina-metrics-extension/src/main/java/org/ballerinalang/observe/metrics/extension/defaultimpl/RollingHistogram.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.ballerinalang.observe.metrics.extension.defaultimpl;
+
+import org.HdrHistogram.DoubleHistogram;
+import org.HdrHistogram.DoubleRecorder;
+import org.ballerinalang.util.metrics.PercentileValue;
+import org.ballerinalang.util.metrics.Snapshot;
+import org.ballerinalang.util.metrics.StatisticConfig;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+/**
+ * Using {@link DoubleHistogram} to maintain samples in a ring buffer to decay older samples and give greater weight
+ * to recent samples. This implementation allows to get summary statistics for a rolling window over the last X minutes.
+ */
+public class RollingHistogram {
+
+    /**
+     * Clock for measuring time.
+     */
+    private final Clock clock;
+
+    /**
+     * Statistic configurations used for this {@link RollingHistogram}.
+     */
+    private final StatisticConfig statisticConfig;
+
+    /**
+     * Recorder in a ring buffer.
+     */
+    private final DoubleRecorder[] ringBuffer;
+
+    /**
+     * Current bucket index.
+     */
+    private int currentBucket;
+
+    /**
+     * Last time the {@link RollingHistogram} was rotated.
+     */
+    private volatile long lastRotateTimestampMillis;
+
+    /**
+     * Duration between rotations.
+     */
+    private final long durationBetweenRotatesMillis;
+
+    /**
+     * This interval histogram is reused when taking the next interval histogram.
+     */
+    private final DoubleHistogram intervalHistogram;
+
+    /**
+     * A histogram to keep accumulated stats.
+     */
+    private DoubleHistogram accumulatedHistogram;
+
+    /**
+     * Global atomic field updater to update volatile {@code rotating} integer.
+     */
+    private static final AtomicIntegerFieldUpdater<RollingHistogram> rotatingUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(RollingHistogram.class, "rotating");
+
+    /**
+     * A flag to check whether {@link RollingHistogram} is being rotated.
+     */
+    private volatile int rotating; // 0 - not rotating, 1 - rotating
```

Comment:
Do we need this. It is not used after the value assignement

Reply:
Yes, this variable was used by AtomicIntegerFieldUpdater-rotatingUpdater during rolling.